haskell/test/SMT/Model/BaseDAO/Proposal.hs

```diff
@@ -0,0 +1,368 @@
+  lvl <- get <&> msLevel
+  pure (lvl >= (proposal & plStartLevel) + target)
+
+handleProposalIsOver :: (Natural, ProposalKey) -> ModelT Bool
+handleProposalIsOver (proposalLvl, proposalKey) = do
+  config <- getConfig
+  proposal <- checkIfProposalExist proposalKey
+
+  isExpired <- isLevelReached proposal =<< (getConfig <&> cProposalExpiredLevel)
+  canBeFlushed <- isLevelReached proposal =<< (getConfig <&> cProposalFlushLevel)
+  if isExpired then
+    throwError EXPIRED_PROPOSAL
+  else if canBeFlushed then do
+    cond <- getStore <&> \s ->
+              doTotalVoteMeetQuorumThreshold proposal s && ((proposal & plUpvotes) > (proposal & plDownvotes))
+
+    unfreezeProposerAndVoterToken cond proposal
+    when cond $ do
+      DecisionLambdaOutput ops newExtra guardianMaybe <- getStore
+        >>= \s -> (config & cDecisionLambda) -$$? (DecisionLambdaInput proposal (s & sExtra))
+
+      case guardianMaybe of
+        Just g -> modifyStore $ \s -> s { sGuardian = g }
+        Nothing -> pure ()
+
+      modifyStore $ \s -> s { sExtra = newExtra }
+      execOperations (mkSimpleOperation <$> ops)
+
+    deleteProposal proposalLvl proposalKey
+    pure True
+  else
+    pure False
+
+applyFlush :: ModelCall -> Natural -> ModelT ()
+applyFlush _ param = do
+  when (param == 0) $
+    throwError BAD_ENTRYPOINT_PARAMETER
+
+  anyFlushedProposal <-
+    getStore
+      <&> (take (fromIntegral param) . Set.toAscList . sProposalKeyListSortByDate)
+      >>= mapM handleProposalIsOver
+      <&> or
+
+  unless anyFlushedProposal $ throwError EMPTY_FLUSH
+
+applyDropProposal :: ModelCall -> ProposalKey -> ModelT ()
+applyDropProposal mc proposalKey = do
+
+  proposal <- checkIfProposalExist proposalKey
+  proposalIsExpired <- isLevelReached proposal =<< (getConfig <&> cProposalExpiredLevel)
+
+  store <- getStore
+  if   ((mc & mcSender) == (proposal & plProposer))
+    || ((mc & mcSender) == (store & sGuardian) && (mc & mcSender) /= (mc & mcSource))
+    || proposalIsExpired
+  then do
+    unfreezeProposerAndVoterToken False proposal
+    deleteProposal (proposal & plStartLevel) proposalKey
+  else
+    throwError DROP_PROPOSAL_CONDITION_NOT_MET
+
+---------------------------------------------------------------
+-- Delegate
+---------------------------------------------------------------
+
+updateDelegate :: ModelCall -> Set Delegate -> DelegateParam -> Set Delegate
+updateDelegate mc delegates param =
+  if param & dpEnable
+    then Set.insert delegate delegates
+    else Set.delete delegate delegates
+  where
+    delegate = Delegate { dOwner = mc & mcSender, dDelegate = param & dpDelegate }
+
+applyUpdateDelegate :: ModelCall -> [DelegateParam] -> ModelT ()
+applyUpdateDelegate mc params = do
+  delegates <- getStore <&> (Map.keysSet . T.unBigMap . sDelegates)
+  let newDelegates = foldl' (updateDelegate mc) delegates params
+  modifyStore $ \s -> s { sDelegates = BigMap $ Map.fromSet (const ()) newDelegates }
```

Comment:
It could all be written inside a  single `modifyStore` use

Reply:
Updated.